clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4669,12 +4669,21 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     if (HasFPGA)
       CmdArgs.push_back("-fsycl-disable-range-rounding");
 
+    const auto &DeviceTriple = getToolChain().getTriple();
+    const auto &DeviceSubArch = DeviceTriple.getSubArch();
     // Enable generation of USM address spaces for FPGA.
     // __ENABLE_USM_ADDR_SPACE__ will be used during compilation of SYCL headers
-    if (getToolChain().getTriple().getSubArch() ==
-        llvm::Triple::SPIRSubArch_fpga)
+    if (DeviceSubArch == llvm::Triple::SPIRSubArch_fpga)
       CmdArgs.push_back("-D__ENABLE_USM_ADDR_SPACE__");
 
+    // Use native FP atomics on supported targets by default.
+    // TODO: Currently, only OpenCL CPU and Intel GPU support the functions.
+    // Allow other targets through check once/if support for
+    // SPV_EXT_shader_atomic_float* extensions lands there.
+    if (DeviceTriple.isSPIR() &&
+        DeviceSubArch != llvm::Triple::SPIRSubArch_fpga)
+      CmdArgs.push_back("-DSYCL_USE_NATIVE_FP_ATOMICS");
```

Comment:
Sorry if this is a silly question, but I don't know how `CmdArgs` works.

After this change, how does a user compile for CPU or GPU and disable the use of SPV_EXT_shader_atomic_float?  I'm thinking about the use-case where somebody is trying to produce SPIR-V without any extensions, to run on a different OpenCL runtime.

Reply:
> After this change, how does a user compile for CPU or GPU and disable the use of SPV_EXT_shader_atomic_float?

That's a good point that I haven't thought about much. The bluntest way would be to pass `-USYCL_USE_NATIVE_FP_ATOMICS`, as is done in the [SYCL LIT tests](https://github.com/intel/llvm/pull/3869/files#diff-1f987d34acf2296eed8833883807005e2874777d72cbf6a972e27eda512aa65f) within this PR. Maybe the driver should also provide ways of configuring SPIR-V extensions that are enabled within the `llvm-spirv` command, but this is a more general item far out of this PR's scope.

Is the approach with explicit undefinition of the macro acceptable? Clearly, it's not too convenient, but neither has been the macro _definition_ for the larger set of use-cases...

Reply:
> Is the approach with explicit undefinition of the macro acceptable? Clearly, it's not too convenient, but neither has been the macro definition for the larger set of use-cases...

Yes, I think that would be acceptable for now.  I mainly wasn't sure whether the user would be able to pass something like an undefine, or whether `CmdArgs` would always take precedence -- thanks for the explanation.

> Maybe the driver should also provide ways of configuring SPIR-V extensions that are enabled within the llvm-spirv command, but this is a more general item far out of this PR's scope.

I agree on both counts.  I recall talking to @AlexeySachkov about this as part of the general handling of extensions in DPC++.  I don't know if we still have plans to move in that direction, but if we do this seems to be covered by your TODO note about checking for SPV_EXT_shader_atomic_float support.

Reply:
> If we do this seems to be covered by your TODO note about checking for SPV_EXT_shader_atomic_float support.

Technically, that comment more so refers to the particular extension support in the underlying targets. I could file a GitHub issue instead so that we don't lose this.

Reply:
Ah, ok.  In that case, I agree that an issue is a good idea.